cpp/include/cudf/io/text/detail/tile_state.hpp

```diff
@@ -18,47 +20,36 @@ enum class scan_tile_status : uint8_t {
 template <typename T>
 struct scan_tile_state_view {
   uint64_t num_tiles;
-  scan_tile_status* tile_status;
+  cuda::atomic<scan_tile_status, cuda::thread_scope_device>* tile_status;
   T* tile_partial;
   T* tile_inclusive;
 
-  __device__ inline void initialize_status(cudf::size_type base_tile_idx,
-                                           cudf::size_type count,
-                                           scan_tile_status status)
+  __device__ inline void set_status(cudf::size_type tile_idx, scan_tile_status status)
   {
-    auto thread_idx = blockIdx.x * blockDim.x + threadIdx.x;
-    if (thread_idx < count) {  //
-      // this is UB if tile_status gets assigned from multiple threads.
-      tile_status[(base_tile_idx + thread_idx) % num_tiles] = status;
-    }
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    tile_status[offset].store(status);
```

Comment:
You probably want to provide an explicit memory order here. It defaults to sequentially consistent which is the strongest memory model and usually not needed.

Reply:
Any reason we shouldn't use `memory_order_relaxed`?